File: nodejs/src/state/NewOrderState.ts
import OrderState, { IOrder } from "./OrderState";
import PreparingState from "./PreparingState";
import CancelledState from "./CancelledState";

class NewOrderState extends OrderState {
    constructor(order: IOrder) { super(order); }

    getName(): string { return "NEW"; }

    next(): void {
        console.log(`📦 Đơn #${this.order.id}: NEW → PREPARING`);
        this.order.setState(new PreparingState(this.order));
    }

    cancel(): void {
        console.log(`❌ Đơn #${this.order.id}: NEW → CANCELLED`);
        this.order.setState(new CancelledState(this.order));
    }

    addItem(item: { getDescription(): string }): boolean {
        console.log(`✅ Đơn #${this.order.id}: Thêm "${item.getDescription()}" thành công`);
        return true;
    }
}

export default NewOrderState;

File: spring/demo/src/main/java/com/example/demo/state/NewOrderState.java
package com.example.demo.state;

import com.example.demo.decorator.Beverage;
import com.example.demo.model.Order;

public class NewOrderState extends OrderState {
    public NewOrderState(Order order) {
        super(order);
    }

    @Override
    public String getName() {
        return "NEW";
    }

    @Override
    public void next() {
        System.out.println("Đơn #" + order.getId() + ": NEW -> PREPARING");
        order.setState(new PreparingState(order));
    }

    @Override
    public void cancel() {
        System.out.println("Đơn #" + order.getId() + ": NEW -> CANCELLED");
        order.setState(new CancelledState(order));
    }

    @Override
    public boolean addItem(Beverage item) {
        System.out.println("Đơn #" + order.getId() + ": Thêm \"" + item.getDescription() + "\" thành công");
        return true;
    }
}
